// @thi.ng/atom/cursor.js
import { isArray, isFunction } from "@thi.ng/checks";
import { illegalArgs, illegalArity } from "@thi.ng/errors";
import { getter, setter } from "@thi.ng/paths";
import { Atom } from "./atom";
import { nextID } from "./idgen";
import { View } from "./view";
/**
 * A cursor provides read/write access to a path location within a
 * nested parent state (Atom or another Cursor).
 *
 * @remarks
 * Cursors behave like Atoms for all practical purposes, i.e. support
 * {@link @thi.ng/api#IDeref.deref}, {@link IReset.reset},
 * {@link ISwap.swap}, {@link @thi.ng/api#IWatch.addWatch} etc. However,
 * when updating a cursor's value, the parent state will be updated at
 * the cursor's path as well (incl. triggering any watches and/or
 * validators) attached to the parent. Likewise, when the parent state
 * is modified externally, the cursor's value will automatically update
 * as well. The update order of cursor's sharing a common parent is
 * undefined, but can be overridden by extending this class with a
 * custom {@link @thi.ng/api#IWatch.notifyWatches} implementation.
 *
 * If creating multiple cursors w/ a shared parent and each cursor
 * configured with a custom ID (provided via config object to ctor),
 * it's the user's responsibility to ensure the given IDs are unique.
 * Cursors are implemented by attaching a watch to the parent and the ID
 * is used to identify each watch.
 *
 * When using the optional validator predicate (also specified via
 * config object to ctor), the cursor's validator MUST NOT conflict with
 * the one assigned to the parent or else both will go out-of-sync.
 * Therefore, when requiring validation and updating values via cursors
 * it's recommended to only specify validators for leaf-level cursors in
 * the hierarchy.
 */
export class Cursor {
    constructor(...args) {
        let parent;
        let lookup;
        let update;
        let validate;
        let opts;
        let id;
        switch (args.length) {
            case 1:
                opts = args[0];
                id = opts.id;
                parent = opts.parent;
                validate = opts.validate;
                if (opts.path) {
                    if (isArray(opts.path) && isFunction(opts.path[0])) {
                        [lookup, update] = opts.path;
                    }
                    else {
                        lookup = getter(opts.path);
                        update = setter(opts.path);
                    }
                }
                else {
                    illegalArgs("missing path config");
                }
                break;
            case 2:
                parent = args[0];
                lookup = getter(args[1]);
                update = setter(args[1]);
                break;
            case 3:
                [parent, lookup, update] = args;
                break;
            default:
                illegalArity(args.length);
        }
        this.parent = parent;
        this.id = id || `cursor-${nextID()}`;
        this.selfUpdate = false;
        if (!lookup || !update) {
            illegalArgs();
        }
        this.local = new Atom(lookup(parent.deref()), validate);
        this.local.addWatch(this.id, (_, prev, curr) => {
            if (prev !== curr) {
                this.selfUpdate = true;
                parent.swap((state) => update(state, curr));
                this.selfUpdate = false;
            }
        });
        parent.addWatch(this.id, (_, prev, curr) => {
            if (!this.selfUpdate) {
                const cval = lookup(curr);
                if (cval !== lookup(prev)) {
                    this.local.reset(cval);
                }
            }
        });
    }
    get value() {
        return this.deref();
    }
    set value(val) {
        this.reset(val);
    }
    deref() {
        return this.local.deref();
    }
    release() {
        this.local.release();
        this.parent.removeWatch(this.id);
        delete this.local;
        delete this.parent;
        return true;
    }
    reset(val) {
        return this.local.reset(val);
    }
    resetIn(path, val) {
        return this.local.resetIn(path, val);
    }
    swap(fn, ...args) {
        return this.local.swap(fn, ...args);
    }
    swapIn(path, fn, ...args) {
        return this.local.swapIn(path, fn, ...args);
    }
    addWatch(id, fn) {
        return this.local.addWatch(id, fn);
    }
    removeWatch(id) {
        return this.local.removeWatch(id);
    }
    /* istanbul ignore next */
    notifyWatches(oldState, newState) {
        return this.local.notifyWatches(oldState, newState);
    }
    addView(path, tx, lazy = true) {
        return new View(this, path, tx, lazy);
    }
}
